# **Data Collection**

## Objectives

* Collect the Premier League player statistics dataset programmatically from Kaggle.
* Store the collected raw datasets in the `data/raw` folder.
* Inspect each dataset to understand its structure and identify any inconsistencies.
* Prepare the datasets for merging into a single processed DataFrame.

## Inputs

* Premier League player statistics dataset collected from Kaggle.

## Outputs

* Raw Premier League player statistics CSV files stored in the `data/raw` folder.
* A single merged Premier League player statistics DataFrame.
* A processed dataset saved to the `data/processed` folder for use in later notebooks.

## Additional Comments

* The raw datasets are collected programmatically before being processed.
* The original raw data is not modified during the processing stages.


---

### Change working directory

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\code\\premier-league-predictor\\premier-league-predictor\\jupyter_notebooks'

In [2]:
os.chdir(r"C:\code\premier-league-predictor\premier-league-predictor")
print("You set a new current directory")

You set a new current directory


In [3]:
current_dir = os.getcwd()
current_dir

'C:\\code\\premier-league-predictor\\premier-league-predictor'

### Import packages

`pandas` is used for loading, inspecting and processing the datasets.

The Kaggle API is used to collect the original Premier League player statistics dataset programmatically from Kaggle.

In [4]:
import pandas as pd


### Collect dataset from Kaggle

The Premier League player statistics dataset is collected programmatically from Kaggle using the Kaggle API and stored in the `data/raw` directory.

This provides a reproducible data collection mechanism before the raw datasets are inspected, processed and merged.

In [5]:
from kaggle.api.kaggle_api_extended import KaggleApi

api = KaggleApi()
api.authenticate()

api.dataset_download_files(
    "krishanthbarkav/english-premier-leagueepl-player-statistics",
    path="data/raw",
    unzip=True
)

Dataset URL: https://www.kaggle.com/datasets/krishanthbarkav/english-premier-leagueepl-player-statistics


---

### Load all Premier League datasets

The Premier League player statistics datasets are loaded from the `data/raw` directory into pandas DataFrames. These datasets represent multiple seasons and will be standardised before being combined into a single dataset.

In [6]:
data_path = "data/raw"

In [7]:
os.listdir(data_path)

['pl_15-16.csv',
 'pl_16-17.csv',
 'pl_17-18.csv',
 'pl_18-19.csv',
 'pl_19-20.csv',
 'pl_20-21.csv',
 'pl_21-22.csv',
 'pl_22-23.csv',
 'pl_23-24.csv']

In [8]:
csv_files = os.listdir(data_path)
csv_files

['pl_15-16.csv',
 'pl_16-17.csv',
 'pl_17-18.csv',
 'pl_18-19.csv',
 'pl_19-20.csv',
 'pl_20-21.csv',
 'pl_21-22.csv',
 'pl_22-23.csv',
 'pl_23-24.csv']

In [9]:
file_name = csv_files[0]
file_name

'pl_15-16.csv'

In [10]:
file_path = os.path.join(data_path, file_name)
file_path

'data/raw\\pl_15-16.csv'

### Inspect a single file first

The first dataset is explored to understand its structure, identify the available features, and detect any issues before processing all seasons.

In [11]:
df = pd.read_csv(file_path)
df.head()

,Unnamed: 0,Name,Position,Appearances,Clean sheets,Goals conceded,Tackles,Tackle success %,Last man tackles,Blocked shots,...,Shooting accuracy %,Big chances missed,Saves,Penalties saved,Punches,High Claims,Catches,Sweeper clearances,Throw outs,Goal Kicks
0,0,Rolando Aarons,Midfielder,10,NaN,NaN,13.0,77%,NaN,0.0,...,50%,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,1,Almen Abdi,Midfielder,32,NaN,NaN,83.0,78%,NaN,10.0,...,26%,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2,Abdul Rahman Baba,Defender,15,2.0,13.0,47.0,83%,0.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,3,Mehdi Abeid,Midfielder,0,NaN,NaN,0.0,0%,NaN,0.0,...,0%,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,4,Tammy Abraham,Forward,2,NaN,NaN,0.0,NaN,NaN,1.0,...,0%,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [12]:
df.shape

(757, 54)

In [13]:
df.columns

Index(['Unnamed: 0', 'Name', 'Position', 'Appearances', 'Clean sheets',
       'Goals conceded', 'Tackles', 'Tackle success %', 'Last man tackles',
       'Blocked shots', 'Interceptions', 'Clearances', 'Headed Clearance',
       'Clearances off line', 'Recoveries', 'Duels won', 'Duels lost',
       'Successful 50/50s', 'Aerial battles won', 'Aerial battles lost',
       'Own goals', 'Errors leading to goal', 'Assists', 'Passes',
       'Passes per match', 'Big chances created', 'Crosses',
       'Cross accuracy %', 'Through balls', 'Accurate long balls',
       'Yellow cards', 'Red cards', 'Fouls', 'Offsides', 'Goals',
       'Headed goals', 'Goals with right foot', 'Goals with left foot',
       'Hit woodwork', 'Goals per match', 'Penalties scored',
       'Freekicks scored', 'Shots', 'Shots on target', 'Shooting accuracy %',
       'Big chances missed', 'Saves', 'Penalties saved', 'Punches',
       'High Claims', 'Catches', 'Sweeper clearances', 'Throw outs',
       'Goal Kicks'],
 

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 757 entries, 0 to 756
Data columns (total 54 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Unnamed: 0              757 non-null    int64  
 1   Name                    757 non-null    object 
 2   Position                757 non-null    object 
 3   Appearances             757 non-null    int64  
 4   Clean sheets            322 non-null    float64
 5   Goals conceded          322 non-null    float64
 6   Tackles                 681 non-null    float64
 7   Tackle success %        514 non-null    object 
 8   Last man tackles        246 non-null    float64
 9   Blocked shots           681 non-null    float64
 10  Interceptions           681 non-null    float64
 11  Clearances              681 non-null    float64
 12  Headed Clearance        681 non-null    float64
 13  Clearances off line     246 non-null    float64
 14  Recoveries              514 non-null    fl

In [15]:
df.isnull().sum()

Unnamed: 0                  0
Name                        0
Position                    0
Appearances                 0
Clean sheets              435
Goals conceded            435
Tackles                    76
Tackle success %          243
Last man tackles          511
Blocked shots              76
Interceptions              76
Clearances                 76
Headed Clearance           76
Clearances off line       511
Recoveries                243
Duels won                 243
Duels lost                243
Successful 50/50s         243
Aerial battles won        243
Aerial battles lost       243
Own goals                 435
Errors leading to goal    167
Assists                     0
Passes                      0
Passes per match            0
Big chances created        76
Crosses                    76
Cross accuracy %          243
Through balls             243
Accurate long balls       167
Yellow cards                0
Red cards                   0
Fouls                       0
Offsides  

### Load all datasets into memory

Now that the structure of a single file has been inspected, all CSV files in `data/raw` are loaded into a list of DataFrames. A `Season` column is added during loading so that each player record retains the season it belongs to before the datasets are merged.

In [16]:
season_data = []

seasons = [
    "2015-16",
    "2016-17",
    "2017-18",
    "2018-19",
    "2019-20",
    "2020-21",
    "2021-22",
    "2022-23",
    "2023-24",
]

In [17]:
csv_files

['pl_15-16.csv',
 'pl_16-17.csv',
 'pl_17-18.csv',
 'pl_18-19.csv',
 'pl_19-20.csv',
 'pl_20-21.csv',
 'pl_21-22.csv',
 'pl_22-23.csv',
 'pl_23-24.csv']

In [18]:
for file_name, season in zip(csv_files, seasons):
    file_path = os.path.join(data_path, file_name)
    df = pd.read_csv(file_path)

    df["Season"] = season

    season_data.append(df)

In [19]:
"Season" in season_data[0].columns

True

In [20]:
len(season_data)

9

In [21]:
type(season_data[0])

pandas.core.frame.DataFrame

In [22]:
season_data[0].shape

(757, 55)

### Check for column inconsistencies across seasons

The datasets are compared to identify any structural differences before merging. This includes checking for missing or additional columns, differences in column names, and inconsistencies in column order across seasons.

In [23]:
for i, season_df in enumerate(season_data):
    print(i, season_df.shape)

0 (757, 55)
1 (839, 55)
2 (793, 55)
3 (884, 55)
4 (974, 55)
5 (862, 55)
6 (989, 54)
7 (1092, 54)
8 (1016, 54)


In [24]:
set(season_data[0].columns) - set(season_data[6].columns)

{'Big chances created', 'Goals conceded', 'Penalties saved', 'Unnamed: 0'}

### Remove unnecessary index column

Some CSV files were exported with the DataFrame index saved as a column called `Unnamed: 0`. This column is not a player statistic and is removed before merging the datasets.

In [25]:
set(season_data[5].columns) - set(season_data[0].columns)

{'Big Chances Created', 'Goals Conceded', 'Penalties Saved'}

### Standardise column names

Although the datasets contained the same features, three column names used different capitalisation between seasons (e.g. `Big chances created` and `Big Chances Created`). These column names were standardised to ensure a consistent schema before merging the datasets.

In [26]:
for i, season_df in enumerate(season_data):
    print(i, list(season_df.columns))

0 ['Unnamed: 0', 'Name', 'Position', 'Appearances', 'Clean sheets', 'Goals conceded', 'Tackles', 'Tackle success %', 'Last man tackles', 'Blocked shots', 'Interceptions', 'Clearances', 'Headed Clearance', 'Clearances off line', 'Recoveries', 'Duels won', 'Duels lost', 'Successful 50/50s', 'Aerial battles won', 'Aerial battles lost', 'Own goals', 'Errors leading to goal', 'Assists', 'Passes', 'Passes per match', 'Big chances created', 'Crosses', 'Cross accuracy %', 'Through balls', 'Accurate long balls', 'Yellow cards', 'Red cards', 'Fouls', 'Offsides', 'Goals', 'Headed goals', 'Goals with right foot', 'Goals with left foot', 'Hit woodwork', 'Goals per match', 'Penalties scored', 'Freekicks scored', 'Shots', 'Shots on target', 'Shooting accuracy %', 'Big chances missed', 'Saves', 'Penalties saved', 'Punches', 'High Claims', 'Catches', 'Sweeper clearances', 'Throw outs', 'Goal Kicks', 'Season']
1 ['Unnamed: 0', 'Name', 'Position', 'Appearances', 'Clean sheets', 'Goals conceded', 'Tackles

In [27]:
for season_df in season_data:
    if 'Unnamed: 0' in season_df.columns:
        season_df.drop(columns=['Unnamed: 0'], inplace=True)

In [28]:
for i, season_df in enumerate(season_data):
    print(i, season_df.shape)

0 (757, 54)
1 (839, 54)
2 (793, 54)
3 (884, 54)
4 (974, 54)
5 (862, 54)
6 (989, 54)
7 (1092, 54)
8 (1016, 54)


In [29]:
for season_df in season_data:
    season_df.rename(
        columns={
            "Big Chances Created": "Big chances created",
            "Goals Conceded": "Goals conceded",
            "Penalties Saved": "Penalties saved",
        },
        inplace=True,
    )

In [30]:
season_data[7].columns

Index(['Name', 'Position', 'Appearances', 'Goals', 'Goals per match',
       'Headed goals', 'Goals with right foot', 'Goals with left foot',
       'Penalties scored', 'Freekicks scored', 'Shots', 'Shots on target',
       'Shooting accuracy %', 'Hit woodwork', 'Big chances missed', 'Assists',
       'Passes', 'Passes per match', 'Big chances created', 'Crosses',
       'Cross accuracy %', 'Through balls', 'Accurate long balls',
       'Yellow cards', 'Red cards', 'Fouls', 'Offsides', 'Tackles',
       'Tackle success %', 'Blocked shots', 'Interceptions', 'Clearances',
       'Headed Clearance', 'Recoveries', 'Duels won', 'Duels lost',
       'Successful 50/50s', 'Aerial battles won', 'Aerial battles lost',
       'Errors leading to goal', 'Saves', 'Penalties saved', 'Punches',
       'High Claims', 'Catches', 'Sweeper clearances', 'Throw outs',
       'Goal Kicks', 'Clean sheets', 'Goals conceded', 'Own goals',
       'Last man tackles', 'Clearances off line', 'Season'],
      dtype=

In [31]:
season_data[0].columns

Index(['Name', 'Position', 'Appearances', 'Clean sheets', 'Goals conceded',
       'Tackles', 'Tackle success %', 'Last man tackles', 'Blocked shots',
       'Interceptions', 'Clearances', 'Headed Clearance',
       'Clearances off line', 'Recoveries', 'Duels won', 'Duels lost',
       'Successful 50/50s', 'Aerial battles won', 'Aerial battles lost',
       'Own goals', 'Errors leading to goal', 'Assists', 'Passes',
       'Passes per match', 'Big chances created', 'Crosses',
       'Cross accuracy %', 'Through balls', 'Accurate long balls',
       'Yellow cards', 'Red cards', 'Fouls', 'Offsides', 'Goals',
       'Headed goals', 'Goals with right foot', 'Goals with left foot',
       'Hit woodwork', 'Goals per match', 'Penalties scored',
       'Freekicks scored', 'Shots', 'Shots on target', 'Shooting accuracy %',
       'Big chances missed', 'Saves', 'Penalties saved', 'Punches',
       'High Claims', 'Catches', 'Sweeper clearances', 'Throw outs',
       'Goal Kicks', 'Season'],
     

### Align column order across all seasons

Although all datasets contain the same columns, one season stores them in a different order. To ensure a consistent dataset structure before merging, the column order is aligned using the first season as the reference.

In [32]:
season_data[0].columns[3]

'Clean sheets'

In [33]:
season_data[7].columns[3]

'Goals'

In [34]:
for i, (col1, col2) in enumerate(zip(season_data[0].columns, season_data[7].columns)):
    if col1 != col2:
        print(f"{i}: {col1} -> {col2}")

3: Clean sheets -> Goals
4: Goals conceded -> Goals per match
5: Tackles -> Headed goals
6: Tackle success % -> Goals with right foot
7: Last man tackles -> Goals with left foot
8: Blocked shots -> Penalties scored
9: Interceptions -> Freekicks scored
10: Clearances -> Shots
11: Headed Clearance -> Shots on target
12: Clearances off line -> Shooting accuracy %
13: Recoveries -> Hit woodwork
14: Duels won -> Big chances missed
15: Duels lost -> Assists
16: Successful 50/50s -> Passes
17: Aerial battles won -> Passes per match
18: Aerial battles lost -> Big chances created
19: Own goals -> Crosses
20: Errors leading to goal -> Cross accuracy %
21: Assists -> Through balls
22: Passes -> Accurate long balls
23: Passes per match -> Yellow cards
24: Big chances created -> Red cards
25: Crosses -> Fouls
26: Cross accuracy % -> Offsides
27: Through balls -> Tackles
28: Accurate long balls -> Tackle success %
29: Yellow cards -> Blocked shots
30: Red cards -> Interceptions
31: Fouls -> Clearanc

In [35]:
season_data[7] = season_data[7][season_data[0].columns]
season_data[7].columns.equals(season_data[0].columns)

True

In [36]:
reference_columns = season_data[0].columns

for index, season_df in enumerate(season_data):
    season_data[index] = season_df[reference_columns]

In [37]:
for index, season_df in enumerate(season_data):
    print(index, season_df.columns.equals(reference_columns))

0 True
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True


### Merge all season datasets

After standardising the dataset structure, the individual season DataFrames are combined into a single DataFrame. This creates one consolidated dataset that will be passed to the data cleaning stage in the following notebook.

In [38]:
all_players_df = pd.concat(season_data, ignore_index=True)

### Verify merged dataset

The merged dataset is checked to confirm that the new `Season` column has been retained alongside the original player statistics. This ensures each observation can be traced back to the season from which it originated.

In [39]:
all_players_df.shape

(8206, 54)

In [40]:
all_players_df.head()

,Name,Position,Appearances,Clean sheets,Goals conceded,Tackles,Tackle success %,Last man tackles,Blocked shots,Interceptions,...,Big chances missed,Saves,Penalties saved,Punches,High Claims,Catches,Sweeper clearances,Throw outs,Goal Kicks,Season
0,Rolando Aarons,Midfielder,10,NaN,NaN,13.0,77%,NaN,0.0,6.0,...,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2015-16
1,Almen Abdi,Midfielder,32,NaN,NaN,83.0,78%,NaN,10.0,32.0,...,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2015-16
2,Abdul Rahman Baba,Defender,15,2.0,13.0,47.0,83%,0.0,1.0,23.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2015-16
3,Mehdi Abeid,Midfielder,0,NaN,NaN,0.0,0%,NaN,0.0,0.0,...,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2015-16
4,Tammy Abraham,Forward,2,NaN,NaN,0.0,NaN,NaN,1.0,0.0,...,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2015-16


In [41]:
all_players_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8206 entries, 0 to 8205
Data columns (total 54 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Name                    8206 non-null   object 
 1   Position                8206 non-null   object 
 2   Appearances             8206 non-null   int64  
 3   Clean sheets            3579 non-null   float64
 4   Goals conceded          3579 non-null   float64
 5   Tackles                 7244 non-null   float64
 6   Tackle success %        5531 non-null   object 
 7   Last man tackles        2617 non-null   float64
 8   Blocked shots           7244 non-null   float64
 9   Interceptions           7244 non-null   float64
 10  Clearances              7244 non-null   float64
 11  Headed Clearance        7244 non-null   float64
 12  Clearances off line     2617 non-null   float64
 13  Recoveries              5531 non-null   float64
 14  Duels won               5531 non-null   

### Save merged dataset

The merged dataset is saved to the `data/processed` directory. This provides a single, consistent dataset that will be loaded by the data cleaning notebook.

In [42]:
output_path = "data/processed/all_players.csv"

all_players_df.to_csv(output_path, index=False)

In [43]:
os.listdir("data/processed")

['all_players.csv',
 'all_players_cleaned.csv',
 'cleaned_players.csv',
 'engineered_players.csv']

---

## Conclusions and Next Steps

### Conclusions

This notebook successfully collected the Premier League player statistics dataset programmatically from Kaggle using the Kaggle API.

The dataset contains player statistics from nine Premier League seasons. Before merging, the datasets were prepared by:

- Removing unnecessary index columns (`Unnamed: 0`).
- Standardising inconsistent column names.
- Aligning column order across all seasons.
- Adding a `Season` column to preserve the season associated with each player record.

The nine datasets were then combined into a single DataFrame containing **8,206 player records** and **54 features**, which was saved to the `data/processed` directory for further analysis.

### Next Steps

The next notebook will focus on data cleaning. This will include:

- Investigating missing values.
- Checking for duplicate records.
- Validating data types.
- Identifying any additional data quality issues.